### PR TITLE
Add argless `#annotations` overload

### LINKED ISSUE #9322: [RFC] Accessing all annotations on an ivar/type/def
I been thinking about https://github.com/crystal-lang/crystal/issues/6984 and thought of another use case that it would be better suited for.

For the serializer shard I been working on, I have a concept of `PropertyMetadata`.  An array of these types is built out via a macro as a way to generically describe a property; such as its type, name, owning class, etc.

I also have a concept of `ExclusionStrategies`, which allow applying custom runtime logic to determine if a given property should be skipped.  A common implementation I could see for this is defining some custom annotation, applying it to a property/class, and then using the presence of the annotation, or values within the annotation, as part of the exclusion strategy logic.

```cr
annotation SomeAnn; end

class User
  include ASR::Serializable  

  @[SomeAnn(priority: 1, privilege: :private)]
  property manager_id : Int32
end
```

However, given Crystal doesn't have a way to access annotations at runtime, nor is there a clean way to generically do it at compile time, I thought of a pretty decent solution.  If there is a method to return all annotations applied to a given ivar/type/method, then I could include this information as part of the `PropertyMetadata`.  For example, `manager_id` could be represented like:

```cr
<PropertyMetadata(Int32, user)
@name="manager_id",
@annotations={
  "SomeAnn" => {
    priority: 1,
    privilege: :private
  }
}>
```

Then the logic within the strategy could check for a specific annotation:
```cr
return false unless ann = metadata.annotations["SomeAnn"]?

return true if ann[:privilege] == :private

...
```

I think it also would play well for future annotation enhancements, such as being able to do `ivar.annotations.select(ValidationAnnotation)` or something along those lines.

WDYT?

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Resolves #9322.  Was pretty simple so just went and did it.

### CODE REVIEW

**Comment by bcardiff** [comment]
I'm not convinced that `name` is the proposed way to check the annotation type.

I'm not convinced that getting all the annotations together is useful. Wouldn't you need to know the type to know what properties and semantics give to the annotation? If so, why the methods that return only the annotation of a given type is not enough? You will query based on the type you want to give semantics to.

I've expressed this before, that querying annotation should have a design cycle regarding how the type hierarchy plays along. Are annotations attached to descendant types? Can we query for exact on inherited annotations? C# reflection API is a good source for inspiration here. Without discussing those things I think that all these reflection methods are experimental.

I thought the issue was about making the annotation information available on runtime.

**Comment by Blacksmoke16** [comment]
> I'm not convinced that `name` is the proposed way to check the annotation type.

This was more so to return the name of the annotation (i.e. whats in between the `@[]`), rather than its type.  Similar to [TypeNode#name](https://crystal-lang.org/api/master/Crystal/Macros/TypeNode.html#name(*,generic_args:BoolLiteral=true):MacroId-instance-method) or [MetaVar#name](https://crystal-lang.org/api/master/Crystal/Macros/MetaVar.html#name:MacroId-instance-method).

> why the methods that return only the annotation of a given type is not enough? You will query based on the type you want to give semantics to.

> I thought the issue was about making the annotation information available on runtime.

The idea behind #9322 (and by extension this PR) was to provide a simple workaround for the lack of runtime reflection by exposing the data required to get a similar effect.  For example, building out a hash (at compile time) of the annotations and data within the annotations, then using this hash at runtime; can look at my example in the original issue for example code.

> I've expressed this before, that querying annotation should have a design cycle regarding how the type hierarchy plays along.

Annotations in general could use some love.  There are plenty of optimizations, improvements, and such that could make them even more useful.  However, I think that can happen independently of this PR.  The API for `#all_annotations` wouldn't change.  The data the gets returned by it could, or additional functionality could be added in the future without changing current semantics.

**Comment by Blacksmoke16** [comment]
Bump.

File: spec/compiler/semantic/annotation_spec.cr [hunk 1]
```diff
@@ -103,6 +137,205 @@ describe "Semantic: annotation" do
...
+      assert_type(%(
+        annotation Foo; end
+        annotation Bar; end
+
+        @[Foo]
+        @[Bar]
+        module Moo
+        end
+
+        {% if Moo.all_annotations.size == 2 && Moo.all_annotations[0].name == Foo.id && Moo.all_annotations[1].name == Bar.id %}
```
Comment (Sija) [inline spec/compiler/semantic/annotation_spec.cr:166]:
```suggestion
        {% if Moo.all_annotations.map(&.name) ==[Foo.id, Bar.id] %}
```

File: spec/compiler/semantic/annotation_spec.cr [hunk 2]
```diff
@@ -103,6 +137,205 @@ describe "Semantic: annotation" do
...
+      assert_type(%(
+        annotation Foo; end
+        annotation Bar; end
+
+        @[Foo]
+        @[Bar]
+        class Moo
+        end
+
+        {% if Moo.all_annotations.size == 2 && Moo.all_annotations[0].name == Foo.id && Moo.all_annotations[1].name == Bar.id %}
```
Comment (Sija) [inline spec/compiler/semantic/annotation_spec.cr:184]:
```suggestion
        {% if Moo.all_annotations.map(&.name) == [Foo.id, Bar.id] %}
```

File: spec/compiler/semantic/annotation_spec.cr [hunk 3]
```diff
@@ -103,6 +137,205 @@ describe "Semantic: annotation" do
...
+      assert_type(%(
+        annotation Foo; end
+        annotation Bar; end
+
+        @[Foo]
+        @[Bar]
+        struct Moo
+        end
+
+        {% if Moo.all_annotations.size == 2 && Moo.all_annotations[0].name == Foo.id && Moo.all_annotations[1].name == Bar.id %}
```
Comment (Sija) [inline spec/compiler/semantic/annotation_spec.cr:202]:
```suggestion
        {% if Moo.all_annotations.map(&.name) == [Foo.id, Bar.id] %}
```

File: spec/compiler/semantic/annotation_spec.cr [hunk 4]
```diff
@@ -103,6 +137,205 @@ describe "Semantic: annotation" do
...
+        annotation Foo; end
+        annotation Bar; end
+
+        @[Foo]
+        @[Bar]
+        enum Moo
+          A = 1
+        end
+
+        {% if Moo.all_annotations.size == 2 && Moo.all_annotations[0].name == Foo.id && Moo.all_annotations[1].name == Bar.id %}
```
Comment (Sija) [inline spec/compiler/semantic/annotation_spec.cr:221]:
```suggestion
        {% if Moo.all_annotations.map(&.name) == [Foo.id, Bar.id] %}
```

File: spec/compiler/semantic/annotation_spec.cr [hunk 5]
```diff
@@ -103,6 +137,205 @@ describe "Semantic: annotation" do
...
+        annotation Foo; end
+        annotation Bar; end
+
+        @[Foo]
+        @[Bar]
+        lib Moo
+          A = 1
+        end
+
+        {% if Moo.all_annotations.size == 2 && Moo.all_annotations[0].name == Foo.id && Moo.all_annotations[1].name == Bar.id %}
```
Comment (Sija) [inline spec/compiler/semantic/annotation_spec.cr:240]:
```suggestion
        {% if Moo.all_annotations.map(&.name) == [Foo.id, Bar.id] %}
```

File: spec/compiler/semantic/annotation_spec.cr [hunk 6]
```diff
@@ -103,6 +137,205 @@ describe "Semantic: annotation" do
...
+
+        @[Foo(1)]
+        @[Bar(2)]
+        class Parent(T)
+        end
+
+        class Child < Parent(Int32)
+        end
+
+        {% if Child.superclass.all_annotations[0][0] == 1 && Child.superclass.all_annotations[1][0] == 2 %}
```
Comment (Sija) [inline spec/compiler/semantic/annotation_spec.cr:330]:
```suggestion
        {% if Child.superclass.all_annotations.map(&.[0]) == [1, 2] %}
```

**Comment by Blacksmoke16** [comment]
Bump again.  Any thoughts on this @RX14 @waj @asterite?

**Comment by Blacksmoke16** [comment]
Bump again 🙈.

**Comment by Blacksmoke16** [comment]
Don't suppose I can get another review on this?

**Comment by asterite** [comment]
But it's not clear that this is what we want, so first we must discuss this instead of reviewing this PR. But it's unlikely this will happen before 1.0. And after 1.0 it's not a breaking change, so...

**Comment by Blacksmoke16** [comment]
`Annotation#name` has since been added via https://github.com/crystal-lang/crystal/pull/10811. I merged in master and updated this PR to only add `#all_annotations`. Assuming that's still something we want to consider adding.

I personally would find this method useful, but if there are other higher level discussions happening related to this (e.g. maybe #9802) feel free to close.

File: src/compiler/crystal/annotatable.cr [hunk 7]
```diff
@@ -19,5 +19,10 @@ module Crystal
     def annotations(annotation_type : AnnotationType) : Array(Annotation)?
       @annotations.try &.[annotation_type]?
     end
+
+    # Returns all annotations on this type, if any, or `nil` otherwise
+    def all_annotations : Array(Annotation)?
```
Comment (asterite) [inline src/compiler/crystal/annotatable.cr:24]: Is there a reason this isn't called just `annotations`?

Comment (Blacksmoke16) [reply src/compiler/crystal/annotatable.cr:24]: That's a good call. Will fit better with the existing annotations methods. I'll have to play around with it as it'll conflict with the `@annotations` property. But should be easy enough to rename that to something else as its internal only.

Comment (asterite) [reply src/compiler/crystal/annotatable.cr:24]: Oh, I didn't notice this was the Crystal side. I was thinking about the API name.

Comment (Blacksmoke16) [reply src/compiler/crystal/annotatable.cr:24]: Still worth doing. It's a small enough change and keeps everything consistent.

Comment (straight-shoota) [reply src/compiler/crystal/annotatable.cr]: Hm, I'm not so sure about this. I think calling the new method `all_annotations` would be better than renaming the property to `applied_annotations` which looks a bit odd.

Comment (straight-shoota) [reply src/compiler/crystal/annotatable.cr]: There's just no real gain from this big renaming everywhere.

Comment (Blacksmoke16) [reply src/compiler/crystal/annotatable.cr]: Actually yea that's a good point. We can still keep the public API method `#annotations`, but just make it `#all_annotations` internally. Does break the consistency of things but not a huge deal and probably a bit more clear anyway. I can deal with that tonight.

**Review by straight-shoota — APPROVED** [review]

**Review by asterite — APPROVED** [review]

**Pull request merged**